File: wrappers/PHP/ODCU-courses.php
<?php

class ODCU_courses{

    private $userName = "YOUR USERID";
    private $password = "YOUR KEY";
    private $baseURL = "https://opendata.concordia.ca/API/v1/course/";

    function query($url){
        $curl = curl_init();

        curl_setopt_array($curl, array(
        CURLOPT_URL => $url,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_ENCODING => "",
        CURLOPT_MAXREDIRS => 10,
        CURLOPT_TIMEOUT => 30,
        CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
        CURLOPT_CUSTOMREQUEST => "GET",
        CURLOPT_SSL_VERIFYHOST => 0,
        CURLOPT_SSL_VERIFYPEER => 0,
        CURLOPT_HEADER => 0,
        CURLOPT_USERPWD => $this->userName . ":" . $this->password
       
       
        ));

        $response = curl_exec($curl);
        $err = curl_error($curl);

        curl_close($curl);

        if ($err) {
            var_dump($err);
            return false;
        } else {
            return $response;
        }
    } //function query(){
    
    function getDescriptions($cid){
        $url = $this->baseURL."description/filter/$cid";
        $result = $this->query($url);
        return $result;
    }

    function getCatalog($sub, $cat, $car){
        $url = $this->baseURL."catalog/filter/$sub/$cat/$car";
        $result = $this->query($url);
        return $result;
    }

    function getSections($sub, $cat){
        $url = $this->baseURL."section/filter/$sub/$cat";
        $result = $this->query($url);
        return $result;
    }

    function getSchedules($cid, $sub, $cat){
        $url = $this->baseURL."schedule/filter/$cid/$sub/$cat";
        $result = $this->query($url);
        return $result;
    }

    function getSessions($car, $trm, $shn){
        $url = $this->baseURL."session/filter/$car/$trm/$shn";
        $result = $this->query($url);
        return $result;
    }

    function getFaculties($fac, $dep){
        $url = $this->baseURL."faculty/filter/$fac/$dep";
        $result = $this->query($url);
        return $result;
    }
}


//make the class available in javacript
if(isset($_GET['action'])){

    $endpoint = new ODCU_courses();
    switch ($_GET['action']){
        case 'description':
            if(isset($_GET['cid'])){
                echo $endpoint->getDescriptions($_GET['cid']);
            }
            break;

        case 'catalog':
            if(isset($_GET['sub']) and isset($_GET['cat']) and isset($_GET['car'])){
                echo $endpoint->getCatalog($_GET['sub'], $_GET['cat'], $_GET['car']);
            }
            break;

        case 'section':
            if(isset($_GET['sub']) and isset($_GET['cat'])){
                echo $endpoint->getSections($_GET['sub'], $_GET['cat']);
            }
            break;

        case 'schedule':
            if(isset($_GET['cid']) and isset($_GET['sub']) and isset($_GET['cat'])){
                echo $endpoint->getSchedules($_GET['cid'], $_GET['sub'], $_GET['cat']);
            }
            break;

        case 'session':
            if(isset($_GET['car']) and isset($_GET['trm']) and isset($_GET['shn'])){
                echo $endpoint->getSessions($_GET['car'], $_GET['trm'], $_GET['shn']);
            }
            break;

        case 'faculty':
             if(isset($_GET['fac']) and isset($_GET['dep'])){
                echo $endpoint->getFaculties($_GET['fac'], $_GET['dep']);
            }
            break;    
    }
}

?>

File: wrappers/Python/ODCU-courses.py
import urllib.request

user = 'YOUR USERID'
key = 'YOUR KEY'

endpoints = {
    'description': ['course_id'],
    'catalog': ['subject', 'catalog', 'career'],
    'section': ['subject', 'catalog'],
    'schedule': ['course_id', 'subject', 'catalog'],
    'session': ['career', 'term', 'session'],
    'faculty': ['faculty', 'department'],
}


def openConnection(username, password):
    '''Set up authentication and open connection'''
    API_url = "https://opendata.concordia.ca"
    password_mgr = urllib.request.HTTPPasswordMgrWithDefaultRealm()
    password_mgr.add_password(None, API_url, username, password)
    handler = urllib.request.HTTPBasicAuthHandler(password_mgr)
    opener = urllib.request.build_opener(handler)
    opener.open(API_url)
    urllib.request.install_opener(opener)


def validateRequest(endpoint, params):
    '''Validate that a valid argument with the correct number of parameters has been supplied.'''
    if endpoint not in endpoints:
        helpInfo = ', '.join(endpoints.keys())
        print(f'Invalid endpoint. Please use one of the following: {helpInfo}')
        return False
    elif len(params) != len(endpoints[endpoint]):
        helpInfo = ', '.join(endpoints[endpoint])
        print(f'Invalid params. {endpoint} requires the following: {helpInfo}')
    else:
        return True


def makeRequest(endpoint, *args):
    '''Make request and return the response after validation'''
    if validateRequest(endpoint, args):
        params = '/'.join(args)
        url = f'https://opendata.concordia.ca/API/v1/course/{endpoint}/filter/{params}'
        with urllib.request.urlopen(url) as req:
            res = req.read()
        return res
    else:
        return False


if __name__ == "__main__":
    openConnection(user, key)
    response = makeRequest('session', '*', '*', '*')
    print(response)
